### PR TITLE
Add unit tests for lseek and fseek and fix related bugs

### PR DESCRIPTION
**Pull request by CamStan**
### Description (Updated)

This adds unit tests for our `lseek()`, `fseek()`, `ftell()`, and `rewind()`wrappers and fixes some related bugs. These mostly had to do with setting errno when invalid offsets are passed.

The `-D_GNU_SOURCE` is needed in `client/src/Makefile.am` in order to use `SEEK_DATA` and `SEEK_HOLE`.

Also adds tests for `SEEK_DATA` and `SEEK_HOLE` within `lseek()`, and adds the fallback support as stated in the man page:

> For  the  purposes of these operations, a hole
       is a sequence of zeros that (normally) has not
       been allocated in the underlying file storage.
       However, a  file  system  is  not  obliged  to
       report  holes,  so  these operations are not a
       guaranteed mechanism for mapping  the  storage
       space actually allocated to a file.  (Further‐
       more, a sequence of zeros  that  actually  has
       been written to the underlying storage may not
       be reported  as  a  hole.)   In  the  simplest
       implementation,  a file system can support the
       operations by making SEEK_HOLE  always  return
       the  offset of the end of the file, and making
       SEEK_DATA always return offset (i.e., even  if
       the  location referred to by offset is a hole,
       it can be considered to consist of  data  that
       is a sequence of zeros).

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted.

### CODE REVIEW

**Comment by CamStan** [comment]
I can add the fallback support for SEEK_DATA and SEEK_HOLE to this PR.

File: client/src/unifyfs-sysio.c [hunk 1]
```diff
@@ -885,23 +885,36 @@ off_t UNIFYFS_WRAP(lseek)(int fd, off_t offset, int whence)
         unifyfs_fd_t* filedesc = unifyfs_get_filedesc_from_fd(fd);
 
         /* get current file position */
-        off_t current_pos = filedesc->pos;
+        off_t logical_eof, current_pos = filedesc->pos;
```
Comment (adammoody) [inline client/src/unifyfs-sysio.c]: Let's split this declaration into two lines for clarity.

**Comment by adammoody** [comment]
Thanks, @CamStan .  Nice find.  We should support the fallback for SEEK_HOLE/DATA at least.  Sounds like those can effective behave similarly to SEEK_END/SEEK_CUR.

With this, do you mind also doing a check of fseek?

**Comment by CamStan** [comment]
> We should support the fallback for SEEK_HOLE/DATA at least. Sounds like those can effective behave similarly to SEEK_END/SEEK_CUR.
> 
> With this, do you mind also doing a check of fseek?

Yup, I can add all that to this PR

File: client/src/unifyfs-stdio.c [hunk 2]
```diff
@@ -1559,7 +1573,12 @@ void UNIFYFS_WRAP(rewind)(FILE* stream)
         /* lookup stream */
         unifyfs_stream_t* s = (unifyfs_stream_t*) stream;
 
-        /* TODO: check that stream is active */
+        /* check that stream is active */
+        int fd = s->fd;
+        if (s->fd < 0) {
+            errno = EBADF;
+            return;
+        }
```
Comment (CamStan) [inline client/src/unifyfs-stdio.c]: This should be good for review now. It's currently in three logically separate commits, but can squash them if need be.

One question though. One of the tests I wrote was making `rewind` fail when called on a non-open stream. I noticed this `TODO` within most of our `stdio` wrappers and did the check in this way, which made the test pass. However, not sure if this is the right approach as I'm assuming there is a reason the check hadn't been done in any of the wrappers yet.

File: client/src/unifyfs-sysio.c [hunk 3]
```diff
@@ -886,22 +886,57 @@ off_t UNIFYFS_WRAP(lseek)(int fd, off_t offset, int whence)
...
-            current_pos = unifyfs_fid_logical_size(fid);
+            logical_eof = unifyfs_fid_logical_size(fid);
+            if (logical_eof + offset < 0) {
+                /* offset is negative and will result in negative position */
+                errno = EINVAL;
+                return (off_t)(-1);
+            }
+            current_pos = logical_eof + offset;
+            break;
+        case SEEK_DATA:
```
Comment (adammoody) [inline client/src/unifyfs-sysio.c]: What does the team think of adding a ```#ifdef _GNU_SOURCE``` to conditionally compile these?  Worth doing that?

File: client/src/unifyfs-stdio.c (hunk 2 above)
Comment (CamStan) [reply client/src/unifyfs-stdio.c]: Noticed I assigned to a local var but didn't use. I'll fix that.

And I verified that `unifyfs_stream_init` does also set `fd` back to -1.
https://github.com/LLNL/UnifyFS/blob/872cb269f23d8e6473ba43032b47e34d5259769b/client/src/unifyfs-stdio.c#L1851-L1852

**Comment by adammoody** [comment]
Nice addition, @CamStan .

I don't see SEEK_HOLE/DATA in the man page for fseek, but are those supported there?  Guessing not, since you didn't add them.

I also realize now that a lot of these lseek/fseek tests will fail upon merging https://github.com/LLNL/UnifyFS/pull/437 due to our lack of implied fsync noted in https://github.com/LLNL/UnifyFS/issues/470.

File: client/src/unifyfs-sysio.c (hunk 3 above)
Comment (CamStan) [reply client/src/unifyfs-sysio.c]: I had that thought as well, and wouldn't be opposed. If so, does this effect how it is added in the Makefile (i.e., does it need to become a config option or anything)?

**Comment by CamStan** [comment]
> I don't see SEEK_HOLE/DATA in the man page for fseek, but are those supported there? Guessing not, since you didn't add them.

I didn't see them in the man page either and so assumed they were not supported in `fseek`. A quick test program (not using UnifyFS) of this seem to confirm that (did this with both SEEK_DATA and SEEK_HOLE):
```
/* continued from tests that already wrote to a file */

/* fseek passed end of file */
ret = fseek(fd, 25, SEEK_END);

printf("ftell after seeking beyond end of file with SEEK_END\n");
pos = ftell(fd);
printf("pos: %ld, errno: %d, strerror: %s\n", pos, errno, strerror(errno));

/* fwrite passed end of file to create a hole */
rc = fwrite("hello universe", 15, 1, fd);
printf("ftell after write\n");
pos = ftell(fd);
printf("pos: %ld, errno: %d, strerror: %s\n", pos, errno, strerror(errno));

/* fseek with SEEK_DATA should go to current offset (15) or offset of first data (37) */
printf("fseek to data passed the hole with SEEK_DATA\n");
ret = fseek(fd, 15, SEEK_DATA);
printf("ret: %d, errno: %d, strerror: %s\n", ret, errno, strerror(errno));

printf("ftell after seeking passed hole with SEEK_DATA\n");
pos = ftell(fd);
printf("pos: %ld, errno: %d, strerror: %s\n", pos, errno, strerror(errno));
```
Results in:
```
ftell after seeking beyond end of file with SEEK_END
pos: 37, errno: 0, strerror: Success
ftell after write
pos: 52, errno: 0, strerror: Success
fseek to data passed the hole with SEEK_DATA
ret: -1, errno: 22, strerror: Invalid argument
ftell after seeking passed hole with SEEK_DATA
pos: 52, errno: 0, strerror: Success
```

> I also realize now that a lot of these lseek/fseek tests will fail upon merging #437 due to our lack of implied fsync noted in #470.

**Comment by CamStan** [comment]
> I don't see SEEK_HOLE/DATA in the man page for fseek, but are those supported there? Guessing not, since you didn't add them.

I didn't see them in the man page either and so assumed they were not supported in `fseek`. A quick test program (not using UnifyFS) of this seem to confirm that they are seen as invalid arguments (did this with both SEEK_DATA and SEEK_HOLE):
```
/* continued from tests that already wrote "hello world" to a file */

printf("fseek to beyond end of file with SEEK_END\n");
ret = fseek(fd, 25, SEEK_END);
printf("ret: %d, errno: %d, strerror: %s\n", ret, errno, strerror(errno));

printf("ftell after seeking beyond end of file with SEEK_END\n");
pos = ftell(fd);
printf("pos: %ld, errno: %d, strerror: %s\n", pos, errno, strerror(errno));

/* fwrite passed end of file to create a hole */
rc = fwrite("hello universe", 15, 1, fd);
printf("ftell after write\n");
pos = ftell(fd);
printf("pos: %ld, errno: %d, strerror: %s\n", pos, errno, strerror(errno));

/* fseek with SEEK_DATA should go to current offset (15) or offset of first data (37) */
printf("fseek to data passed the hole with SEEK_DATA\n");
ret = fseek(fd, 15, SEEK_DATA);
printf("ret: %d, errno: %d, strerror: %s\n", ret, errno, strerror(errno));

printf("ftell after seeking passed hole with SEEK_DATA\n");
pos = ftell(fd);
printf("pos: %ld, errno: %d, strerror: %s\n", pos, errno, strerror(errno));
```
Results in:
```
fseek to beyond end of file with SEEK_END
ret: 0, errno: 0, strerror: Success
ftell after seeking beyond end of file with SEEK_END
pos: 37, errno: 0, strerror: Success
ftell after write
pos: 52, errno: 0, strerror: Success
fseek to data passed the hole with SEEK_DATA
ret: -1, errno: 22, strerror: Invalid argument
ftell after seeking passed hole with SEEK_DATA
pos: 52, errno: 0, strerror: Success
```

---

> I also realize now that a lot of these lseek/fseek tests will fail upon merging #437 due to our lack of implied fsync noted in #470.

Is this because #437 changes `unifyfs_fid_logical_size(int fid)` to call `invoke_client_filesize_rpc()`?
If I just put an `fsync()` after the writes in these tests, will that fix it?

**Comment by adammoody** [comment]
Thanks for double checking the question about fseek.  That's good to know.

Yes, I think adding an fsync() after the write() calls in the new tests here would fix the problem.  However, let's hold off on doing that.  First, we can look into what it would take to fix things so we can avoid those fsync() calls.  It may be enough to check the dirty bit and sync if needed before invoking the filesize rpc.  If things get too complicated with that approach, we can come back to add fsync() in the tests here.

**Comment by adammoody** [comment]
Will go ahead and merge this, but let's still bring up the question about _GNU_SOURCE in our next group meeting.

**Comment by CamStan** [comment]
> ... let's still bring up the question about _GNU_SOURCE in our next group meeting.

Will do

**Pull request merged**